### PR TITLE
CLI Metadata Foundations

### PR DESCRIPTION
**Pull request by shauns**
### WHY are these changes introduced?

To answer:
1. How to gather data from plug-ins when from centralised functionality like metadata logging
2. How to track transient runtime metadata during a command's execution
3. How to be extremely careful about sensitive & public data re: Shopify Monorail

This _doesn't_ include tracking of any new data, and doesn't touch Bugsnag metadata.

### WHAT is this pull request doing?

- changes how reporting of Shopify Monorail events is performed
- adds hooks for capturing plug-in wide metadata
- adds a container for runtime metadata

### How to test your changes?

`yarn shopify app info --verbose --path=<your project>` , will include the analytics payload (but won't submit it) in the output.

### CODE REVIEW

File: packages/cli-kit/src/analytics.ts [hunk 1]
```diff
@@ -87,24 +91,37 @@ const buildPayload = async (errorMessage: string | undefined, currentTime: numbe
...
+    project_type: await getProjectType(joinPath(directory, 'web')),
+    api_key: appInfo?.appId,
+    partner_id: partnerIdAsInt,
+  }
+
   return {
-    schema_id: 'app_cli3_command/1.0',
-    payload: {
-      project_type: await getProjectType(joinPath(directory, 'web')),
+    public: {
```
Comment (isaacroldan) [inline packages/cli-kit/src/analytics.ts:104]: are we changing the payload schema? is this OK with monorail?  (asking without knowing how monorail events work)

Comment (shauns) [reply packages/cli-kit/src/analytics.ts:104]: It ends up being the same, see `monorail.ts.Schemas`

File: packages/create-app/src/commands/init.test.ts [hunk 2]
```diff
@@ -46,14 +47,14 @@ describe('create app command', () => {
     // When
     await Init.run(['--template', 'java'])
 
-    expect(errorHandler).toHaveBeenCalledWith(UnsupportedTemplateAlias())
+    expect(errorHandler).toHaveBeenCalledWith(UnsupportedTemplateAlias(), expect.any(Config))
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/create-app/src/commands/init.test.ts:50]: ah! I see what you mean, i've actually used this same feature before 🤦

File: packages/cli-kit/src/node/error-handler.ts [hunk 3]
```diff
@@ -26,15 +27,17 @@ export function errorHandler(error: Error & {exitCode?: number | undefined}) {
...
   }
 }
 
-const reportError = async (error: Error): Promise<Error> => {
-  await reportEvent({errorMessage: error.message})
+const reportError = async (error: Error, config?: Interfaces.Config): Promise<Error> => {
+  if (config !== undefined) {
+    await reportEvent({config, errorMessage: error.message})
+  }
   if (settings.debug || !shouldReportError(error)) return error
```
[the remark this answers is not in the document]
Comment (shauns) [reply packages/cli-kit/src/node/error-handler.ts:41]: The `runCLI` uses `catch`, and connects it to `errorHandler`, which connects to this, but there's no command config available at that point, as it's happening outside of the scope of a command. I don't really want to _drop_ the `catch` because there may be some benefit or reason to it and it does still hit Bugsnag.

**Review by isaacroldan — APPROVED** [review]

**Review by gonzaloriestra — APPROVED** [review]
Nice work! Just a few thoughts:
- Some parts are a bit hard to follow, like some new complex interfaces and types (especially for a TS beginner 😅), but not sure if there's a simpler way to do it...
- The plugins part looks like a good idea, but in general I prefer to avoid adding new code until we really need it. Or are we planning to add metadata for any plugin soon?

File: packages/cli-kit/src/node/error-handler.ts [hunk 4]
```diff
@@ -26,15 +27,17 @@ export function errorHandler(error: Error & {exitCode?: number | undefined}) {
...
       .then(() => {
         process.exit(1)
       })
   }
 }
 
-const reportError = async (error: Error): Promise<Error> => {
-  await reportEvent({errorMessage: error.message})
+const reportError = async (error: Error, config?: Interfaces.Config): Promise<Error> => {
+  if (config !== undefined) {
```
[the remark this answers is not in the document]
Comment (shauns) [reply packages/cli-kit/src/node/error-handler.ts:38]: Yes - Isaac raised the same point, but we attach the error handler _before_ the command is instantiated. We also attach it to the command as well, and this is what most errors end up being caught by.

**Comment by shauns** [comment]
> * planning to add metadata for any plugin soon?

Yep! This PR would get too big with the instrumentation added though I think.
